igvc/src/global_mapper/main.cpp

```diff
@@ -10,57 +10,90 @@
 #include <ros/ros.h>
 #include <stdlib.h>
 #include <string>
+#include <queue>
 
-pcl::PointCloud<pcl::PointXYZ>::Ptr global_map;
+pcl::PointCloud<pcl::PointXYZRGB>::Ptr global_map;
 ros::Publisher _pointcloud_pub;
 ros::Publisher _pointcloud_incremental_pub;
+ros::Publisher _pointcloud_prob_pub;
 tf::TransformListener *tf_listener;
 std::set<std::string> frames_seen;
 bool firstFrame;
 double maxCorrDist;
 int maxIter;
-pcl::octree::OctreePointCloudSearch<pcl::PointXYZ> octree(.01f);
+pcl::octree::OctreePointCloudSearch<pcl::PointXYZRGB>::Ptr octree;
 double searchRadius;
+double octree_resolution;
+double probability_thresh;
 
-void icp_transform(pcl::PointCloud<pcl::PointXYZ>::Ptr input)
+//Probabilistic mapping
+std::vector<pcl::PointCloud<pcl::PointXYZRGB>::Ptr> pointcloud_list;
+int queue_window_size;
+
+void icp_transform(pcl::PointCloud<pcl::PointXYZRGB>::Ptr input)
 {
-  if (!firstFrame)
+  if (pointcloud_list.size() < queue_window_size)
+  {
+    pointcloud_list.push_back(input);
+  } 
+  if (pointcloud_list.size() == queue_window_size)
   {
-    pcl::IterativeClosestPoint<pcl::PointXYZ, pcl::PointXYZ> icp;
-    icp.setMaxCorrespondenceDistance(maxCorrDist);
-    icp.setMaximumIterations(maxIter);
-    icp.setInputSource(input);
-    icp.setInputTarget(global_map);
-    pcl::PointCloud<pcl::PointXYZ> Final;
-    pcl::PointCloud<pcl::PointXYZ> new_points;
-    icp.align(Final);
-    for (unsigned int i = 0; i < Final.size(); ++i)
+    //Probabilistic calculations here
+    pcl::PointCloud<pcl::PointXYZRGB>::Ptr local_pc = pcl::PointCloud<pcl::PointXYZRGB>::Ptr(new pcl::PointCloud<pcl::PointXYZRGB>());
+    pcl::octree::OctreePointCloudSearch<pcl::PointXYZRGB> local_octree(octree_resolution);
+
+    local_octree.setInputCloud(local_pc);
+    for (unsigned int i = 0; i < pointcloud_list.size(); i++)
     {
-      pcl::PointXYZ searchPoint(Final[i].x, Final[i].y, Final[i].z);
-      std::vector<int> indices;
-      std::vector<float> distances;
-      if (octree.radiusSearch(searchPoint, searchRadius, indices, distances, 1) == 0)
+      for (auto point : (*pointcloud_list[i]))
       {
-        octree.addPointToCloud(searchPoint, global_map);
-        new_points.push_back(searchPoint);
+        //G and B represent the probability, they are the same value
+        //Add a point to the local octree if we haven't seen it before
+        if (!local_octree.isVoxelOccupiedAtPoint(point))
+        {
+          point.r = 255;
+          point.g = 0;
+          point.b = 0;
+          local_octree.addPointToCloud(point, local_pc);
+        } else
+        {
```

Comment:
Is this inline with the ROS guidelines?

Reply:
Do you mean brackets for conditionals? I'll check

Reply:
the bracket is consistent with ROS CppStyleGuide, if that is what you mean.

Reply:
That is weird but okay